providers/websockets/bitfinex/messages.go

```diff
@@ -0,0 +1,157 @@
+	// EventError indicates that an error occurred.
+	EventError Event = "error"
+	// ChannelTicker is the channel name for the ticker channel.
+	ChannelTicker Channel = "ticker"
+	// IDHeartbeat is the id always used for a heartbeat.
+	IDHeartbeat = "hb"
+	// ExpectedStreamPayloadLength is the expected length of the payload of a data stream.
+	ExpectedStreamPayloadLength = 10
+	// ExpectedBaseStreamLength is the expected length of a stream base message.
+	ExpectedBaseStreamLength = 2
+)
+
+type ErrorCode int64
+
+const (
+	ErrorUnknownEvent       ErrorCode = 10000
+	ErrorUnknownPair        ErrorCode = 10001
+	ErrorLimitOpenChannels  ErrorCode = 10305
+	ErrorSubscriptionFailed ErrorCode = 10400
+	ErrorNotSubscribed      ErrorCode = 104001
+)
+
+// Error returns the string representation of the ErrorCode.
+func (e ErrorCode) Error() error {
+	switch e {
+	case ErrorUnknownEvent:
+		return errors.New("unknown event")
+	case ErrorUnknownPair:
+		return errors.New("unknown pair")
+	case ErrorLimitOpenChannels:
+		return errors.New("limit of open channels reached")
+	case ErrorSubscriptionFailed:
+		return errors.New("subscribed failed")
+	case ErrorNotSubscribed:
+		return errors.New("not subscribed")
+	default:
+		return errors.New("unknown error type")
+
+	}
+}
+
+// BaseMessage is the base message structure for subscription requests and responses in the
+// BitFinex websocket API.
+type BaseMessage struct {
+	Event string `json:"event" validate:"required"`
+}
+
+// SubscribeMessage is a base message used to make a subscription request.
+//
+// Ex:
+//
+//	{
+//	 event: "subscribe",
+//	 channel: "ticker",
+//	 symbol: SYMBOL
+//	}
+//
+// ref: https://docs.bitfinex.com/reference/ws-public-ticker
+type SubscribeMessage struct {
+	BaseMessage
+	Channel string `json:"channel" validate:"required"`
+	Symbol  string `json:"symbol" validate:"required"`
+}
+
+// NewSubscribeMessages creates subscription messages for the given tickers.
+func NewSubscribeMessages(symbols []string) ([]handlers.WebsocketEncodedMessage, error) {
+	var msgs []handlers.WebsocketEncodedMessage
+
+	if len(symbols) == 0 {
+		return nil, fmt.Errorf("symbols cannot be empty")
+	}
+
+	for _, symbol := range symbols {
+		msg, err := NewSubscribeMessage(symbol)
+		if err != nil {
+			return nil, fmt.Errorf("error marshalling subscription message: %w", err)
+		}
+
+		msgs = append(msgs, msg)
+	}
+
+	return msgs, nil
+}
+
+// NewSubscribeMessage creates a new subscribe message given the ticker symbol
+func NewSubscribeMessage(symbol string) (handlers.WebsocketEncodedMessage, error) {
+	return json.Marshal(SubscribeMessage{
+		BaseMessage: BaseMessage{
+			Event: string(EventSubscribe),
+		},
+		Channel: string(ChannelTicker),
+		Symbol:  symbol,
+	})
+}
+
+// SubscribedMessage is message indicating the status of a subscription request.
+//
+// Ex:
+//
+//	{
+//	  event: "subscribed",
+//	  channel: "ticker",
+//	  chanId: CHANNEL_ID,
+//	  symbol: SYMBOL,
```

Comment:
whats the difference between pair and symbol?

Reply:
A symbol will be `tBTCUSD` indicating a ticker for a spot market.  `fBTCUSD` indicates a future market.

the pair is just `BTCUSD`.